capabilities/istio.ts

```diff
@@ -1,15 +1,115 @@
-import { Capability, a } from "pepr";
-
+import { Capability, RegisterKind, a } from "pepr";
+import { K8sAPI } from "./lib/kubernetes-api";
+import { VirtualService } from "@kubernetes-models/istio/networking.istio.io/v1beta1";
 export const Istio = new Capability({
   name: "istio",
   description: "Istio service mesh capability.",
   namespaces: [],
 });
 
-// Use the 'When' function to create a new Capability Action
 const { When } = Istio;
 
-// When(a.<Kind>).Is<Event>().Then(change => change.<changes>
-When(a.Service)
-  .IsCreated()
-  .Then(svc => svc.SetLabel("pepr", "hi-istio"));
+export class IstioVirtualService extends a.GenericKind {
+  spec: VirtualService["spec"];
+  status?: VirtualService["status"];
+  static apiVersion: VirtualService["apiVersion"];
+  static kind: VirtualService["kind"];
+}
+
+RegisterKind(IstioVirtualService, {
+  group: VirtualService.apiVersion.split("/")[0],
+  version: VirtualService.apiVersion.split("/")[1],
+  kind: VirtualService.kind,
+});
+
+// This capability will only create a gateway if:
+// 1. the ingress class's format for namespace is not set (or set to the same namespace as the ingress)
+// 2. the ingress class must not have wildcards in it. This will be SNI and needs to be a complete match
+//
+// If the ingress class's hostname does have a wildcard in it, and the namespace is the same as the ingress, this is an error
+function parseIngressClass(
+  mynamespace: string,
+  input: string
+): { createGateway: boolean; gateway: string } {
+  const regex = /^pepr-(?:(?<namespace>.+?)\.)?(?<gatewayName>.+)$/;
+  const match = input.match(regex);
+
+  if (match) {
+    const { namespace, gatewayName } = match.groups;
+    if (namespace === undefined || namespace === mynamespace) {
+      return {
+        createGateway: true,
+        gateway: gatewayName,
+      };
+    } else {
+      return {
+        createGateway: false,
+        gateway: `${namespace}/${gatewayName}`,
+      };
+    }
+  }
+
+  return {
+    createGateway: false,
+    gateway: undefined,
+  };
+}
+
+// TODO: do we create a custom resource for pepr for the "state store"
```

Comment:
To clarify, it's "put the state on a resource that the user doesn't control", especially if the user could otherwise abuse this to trick Pepr into doing something like deleting other apps' resources.

A custom resource isn't strictly required. Maybe a secret in a "pepr" namespace or something, not sure.

Reply:
this is why I want ownerreferences. the way it is in right now is a placeholder, this PR I'm going to close, because I have it fully working in another branch.